reconcile/utils/saasherder/saasherder.py

```diff
@@ -1051,6 +1055,37 @@ def _process_template(
             )
         return resources, html_url, target_promotion
 
+    def _assemble_channels(
+        self, channel_names: Optional[Iterable[str]]
+    ) -> list[Channel]:
+        """
+        We need to assemble all publisher_uids that are publishing to a channel.
+        These uids are required to validate correctness of promotions.
+        """
+        channel_names_set = set(channel_names or [])
+        channel_map: dict[str, Channel] = {}
+        if not self._all_saas_files:
+            raise RuntimeError("Must set self._all_saas_files at saasherder init")
+        for saas_file in self._all_saas_files:
+            for tmpl in saas_file.resource_templates:
+                for target in tmpl.targets:
+                    if not target.promotion:
+                        continue
+                    for publish in target.promotion.publish or []:
+                        if publish not in channel_names_set:
+                            continue
+                        publisher_uid = target.uid(
+                            parent_saas_file_name=saas_file.name,
+                            parent_resource_template_name=tmpl.name,
+                        )
+                        if publish not in channel_map:
+                            channel_map[publish] = Channel(
+                                name=publish,
+                                publisher_uids=[],
+                            )
+                        channel_map[publish].publisher_uids.append(publisher_uid)
+        return list(channel_map.values())
```

Comment:
Note this is will iterate on all saas files for every `_process_template`, I recommend to build a graph `publisher -> channel -> subscriber` once before hand, then just lookup info when needed.

Btw, to simplify exist check, can use a `defaultdict`:

```python
        channel_map = defaultdict(list)
        for saas_file in self._all_saas_files:
            for tmpl in saas_file.resource_templates:
                for target in tmpl.targets:
                    if not target.promotion:
                        continue
                    for publish in target.promotion.publish or []:
                        if publish not in channel_names_set:
                            continue
                        publisher_uid = target.uid(
                            parent_saas_file_name=saas_file.name,
                            parent_resource_template_name=tmpl.name,
                        )
                        channel_map[publish].append(publisher_uid)
        return [Channel(name=k, publisher_uids=v) for k, v in channel_map.items()]
```

Reply:
Very nice catch! Thank you!
Running assemble_channel now only once at init :+1: